src/engine/Service.cpp

```diff
@@ -92,6 +95,44 @@ ResultTable Service::computeResult() {
   serviceIriString.remove_suffix(1);
   ad_utility::httpUtils::Url serviceUrl{serviceIriString};
 
+  if (siblingTree_ != nullptr) {
+    // Get the result of the siblingTree, to (potentially)
+    // reduce complexity of the SERVICE query.
+    auto siblingResult = siblingTree_->getResult();
+
+    const size_t rowLimit = 100;
+    if (siblingResult->size() < rowLimit) {
+      auto siblingVariables = siblingTree_->getVariableColumns();
+
+      // Build value clause for each common variable.
+      std::string valueClauses = "{ ";
+      for (const auto& lVar : parsedServiceClause_.visibleVariables_) {
+        auto it = siblingVariables.find(lVar);
+        if (it == siblingVariables.end()) {
+          continue;
+        }
+        const auto& sVar = *it;
+
+        valueClauses += "VALUES " + sVar.first.name() + " { ";
+        for (size_t rowIndex = 0; rowIndex < siblingResult->size();
+             ++rowIndex) {
+          const auto& optionalString =
+              ExportQueryExecutionTrees::idToStringAndType(
+                  siblingTree_->getRootOperation()->getIndex(),
+                  siblingResult->idTable()(rowIndex, sVar.second.columnIndex_),
+                  siblingResult->localVocab());
```

Comment:
This loop can also be refactored for readability, but we can do this once the other issues are resolved.